File: backend/sdk/handlers/onboarding_tools.py
"""
Onboarding tools for TRPG world initialization.

This module defines MCP tools that Onboarding_Manager uses during the onboarding phase:
- draft_world: Lightweight world draft (genre, theme, lore summary) to unblock sub-agents
- persist_world: Comprehensive persistence (full lore + stats + player state)
- complete: Transition world from 'onboarding' to 'active' phase

Sub-agent tool mappings (persist tools by MCP server):
- Onboarding MCP server:
  - draft_world: Used by Onboarding_Manager directly
  - persist_world: Used by Onboarding_Manager directly
- Subagents MCP server (shared by Action Manager and Onboarding Manager):
  - item_designer: Uses mcp__subagents__persist_item
  - character_designer: Uses mcp__subagents__persist_character_design
  - location_designer: Uses mcp__subagents__persist_location_design
- Action Manager direct tools:
  - change_stat: Used directly by Action Manager

Uses WorldService for filesystem-primary storage.
"""

import logging
import uuid
from datetime import datetime
from typing import Any

from claude_agent_sdk import tool
from services.world_service import WorldService
from sqlalchemy.ext.asyncio import AsyncSession

from sdk.handlers.context import ToolContext
from sdk.loaders import get_tool_description, get_tool_response, is_tool_enabled
from sdk.tools.onboarding import (
    CompleteOnboardingInput,
    DraftWorldInput,
    PersistWorldInput,
    ReadLoreGuidelinesInput,
)

logger = logging.getLogger("OnboardingTools")


# =============================================================================
# Tool Name Constants
# =============================================================================

# Tool names for each sub-agent type (persist tools)
# Sub-agents use the shared "subagents" MCP server tools (not action_manager)
# This server is available to both Action Manager and Onboarding Manager
SUBAGENT_TOOL_NAMES = {
    "item_designer": "mcp__subagents__persist_item",
    "character_designer": "mcp__subagents__persist_character_design",
    "location_designer": "mcp__subagents__persist_location_design",
}


async def _add_world_agents_to_initial_location(
    db: AsyncSession,
    world_id: int,
    world_name: str,
    initial_location_name: str,
) -> int:
    """
    Add all world-specific agents to the initial location's room.

    This is called after world seed generation to place NPCs created during
    onboarding into the starting location.

    Args:
        db: Database session
        world_id: World ID
        world_name: World name for filesystem lookup
        initial_location_name: Name of the initial location

    Returns:
        Number of agents added to the location
    """
    import crud
    from services.agent_filesystem_service import AgentFilesystemService

    # Get list of agent names from world's agents folder
    world_agent_names = AgentFilesystemService.list_world_agents(world_name)
    if not world_agent_names:
        logger.info("No world agents to add to initial location")
        return 0

    # Get the initial location from DB
    initial_location = await crud.get_location_by_name(db, world_id, initial_location_name)
    if not initial_location:
        logger.warning(f"Initial location '{initial_location_name}' not found in database")
        return 0

    added_count = 0
    for agent_name in world_agent_names:
        try:
            # Get agent from database
            agent = await crud.get_agent_by_name(db, agent_name)
            if agent:
                # Add to initial location
                await crud.add_character_to_location(db, agent.id, initial_location.id)
                logger.info(f"✨ Added '{agent_name}' to initial location '{initial_location_name}'")
                added_count += 1
            else:
                logger.warning(f"Agent '{agent_name}' not found in database")
        except Exception as e:
            logger.warning(f"Failed to add agent '{agent_name}' to location: {e}")

    logger.info(f"Added {added_count} agents to initial location")
    return added_count


def generate_default_world_name() -> str:
    """
    Generate a default world name if one is not provided.

    Uses timestamp and short UUID to create a unique name.

    Returns:
        Generated world name (e.g., 'world_20251206_a1b2c3')
    """
    timestamp = datetime.now().strftime("%Y%m%d")
    short_uuid = uuid.uuid4().hex[:6]
    return f"world_{timestamp}_{short_uuid}"


def create_onboarding_tools(ctx: ToolContext) -> list:
    """
    Create onboarding tools (read_lore_guidelines, draft_world, persist_world, complete).

    Args:
        ctx: Unified tool context containing agent info and dependencies

    Returns:
        List of onboarding tool functions configured with agent name
    """
    tools = []

    # ==========================================================================
    # read_lore_guidelines tool - Read-only reference for lore writing
    # ==========================================================================
    if is_tool_enabled("read_lore_guidelines"):
        tools.append(_create_read_lore_guidelines_tool(ctx))

    # ==========================================================================
    # draft_world tool - Lightweight world draft to unblock sub-agents
    # Called FIRST, before sub-agents start
    # ==========================================================================
    tools.append(_create_draft_world_tool(ctx))

    # ==========================================================================
    # persist_world tool - Comprehensive persistence (full lore + stats)
    # Called AFTER sub-agents have started with draft context
    # ==========================================================================
    tools.append(_create_persist_world_tool(ctx))

    # ==========================================================================
    # Complete tool - Lightweight phase transition
    # Called LAST after draft_world, persist_world, and sub-agents
    # ==========================================================================
    if is_tool_enabled("complete"):
        complete_description = get_tool_description("complete", agent_name=ctx.agent_name, group_name=ctx.group_name)

        @tool("complete", complete_description, CompleteOnboardingInput.model_json_schema())
        async def complete_tool(args: dict[str, Any]):
            """
            Finalize world transition from 'onboarding' to 'active' phase.

            This tool:
            1. Sets the player name in world config
            2. Sets pending_phase to 'active' for phase transition
            3. Captures initial state (stats, inventory, location) for reset functionality

            Call this AFTER:
            - draft_world (genre, theme, lore summary)
            - Sub-agents (location_designer, character_designer, item_designer)
            - persist_world (full lore, stats)
            """
            from services.location_service import LocationService
            from services.player_service import PlayerService
            from services.world_reset_service import WorldResetService

            validated = CompleteOnboardingInput(**args)
            player_name = validated.player_name
            starting_location = validated.starting_location
            starting_hour = validated.starting_hour

            logger.info(f"World completion requested by {ctx.agent_name}")
            logger.info(
                f"Player: {player_name}, Starting location: {starting_location}, Starting hour: {starting_hour}"
            )

            effective_world_name = ctx.world_name
            if not effective_world_name:
                effective_world_name = generate_default_world_name()
                logger.info(f"Generated default world name: {effective_world_name}")

            try:
                # Validate that starting_location exists in filesystem
                fs_locations = LocationService.load_all_locations(effective_world_name)
                if starting_location not in fs_locations:
                    available = ", ".join(fs_locations.keys()) if fs_locations else "none"
                    return {
                        "content": [
                            {
                                "type": "text",
                                "text": f"Starting location '{starting_location}' not found. "
                                f"Available locations: {available}. "
                                "Make sure location_designer created this location first.",
                            }
                        ],
                        "is_error": True,
                    }

                # Ensure world directory exists
                config = WorldService.ensure_world_exists(effective_world_name)

                # Set player name and transition phase
                config.user_name = player_name
                config.pending_phase = "active"
                WorldService.save_world_config(effective_world_name, config)

                logger.info(f"✅ World '{effective_world_name}' pending phase set to 'active'")

                # Get player state and set the starting location
                player_state = PlayerService.load_player_state(effective_world_name)
                if player_state:
                    player_state.current_location = starting_location
                    player_state.game_time = {"hour": starting_hour, "minute": 0, "day": 1}
                    PlayerService.save_player_state(effective_world_name, player_state)
                    logger.info(f"Set starting location to {starting_location}, game time to {starting_hour}:00")

                # Capture initial state for reset functionality
                # This includes inventory created by item_designer during onboarding
                if player_state:
                    initial_state = WorldResetService.create_initial_state_snapshot(
                        starting_location=starting_location,
                        initial_stats=player_state.stats,
                        initial_inventory=player_state.inventory,
                        initial_game_time=player_state.game_time,
                    )
                    WorldResetService.save_initial_state(effective_world_name, initial_state)
                    logger.info(
                        f"Captured initial state: location={starting_location}, {len(player_state.stats)} stats, {len(player_state.inventory)} items"
                    )

                # Add any world agents (NPCs created during onboarding) to the initial location
                if ctx.db is not None and ctx.world_id is not None:
                    try:
                        await _add_world_agents_to_initial_location(
                            db=ctx.db,
                            world_id=ctx.world_id,
                            world_name=effective_world_name,
                            initial_location_name=starting_location,
                        )
                    except Exception as agent_err:
                        logger.warning(f"Failed to add agents to initial location: {agent_err}")

            except Exception as e:
                logger.error(f"Failed to complete onboarding: {e}", exc_info=True)
                return {
                    "content": [{"type": "text", "text": f"Error completing onboarding: {e}"}],
                    "is_error": True,
                }

            # Get response template from config
            response_text = get_tool_response(
                "complete",
                group_name=ctx.group_name,
                player_name=player_name,
                starting_location=starting_location,
                starting_hour=starting_hour,
            )

            return {"content": [{"type": "text", "text": response_text}]}

        tools.append(complete_tool)

    return tools


# =============================================================================
# Read Lore Guidelines Tool (Read-only reference)
# =============================================================================


def _create_read_lore_guidelines_tool(ctx: ToolContext):
    """
    Create the read_lore_guidelines tool for Onboarding_Manager.

    This tool returns the lore writing guidelines as a reference
    for creating world lore during onboarding.

    Args:
        ctx: Tool context with agent info

    Returns:
        Tool function for reading lore guidelines
    """
    description = get_tool_description("read_lore_guidelines", agent_name=ctx.agent_name, group_name=ctx.group_name)
    # Load lore guidelines content from YAML (via special handler in get_tool_description)
    lore_guidelines_content = get_tool_description("lore_guidelines") or ""

    @tool(
        "read_lore_guidelines",
        description,
        ReadLoreGuidelinesInput.model_json_schema(),
    )
    async def read_lore_guidelines(_args: dict[str, Any]):
        """Return the lore writing guidelines for world creation."""
        ReadLoreGuidelinesInput()  # Validate input (no-op)
        return {"content": [{"type": "text", "text": lore_guidelines_content}]}

    return read_lore_guidelines


# =============================================================================
# Draft World Tool (Lightweight - unblocks sub-agents)
# =============================================================================


def _create_draft_world_tool(ctx: ToolContext):
    """
    Create the draft_world tool for Onboarding_Manager.

    This tool creates a lightweight world draft with genre, theme, and a
    one-paragraph lore summary. Sub-agents can start immediately with this
    context while OM refines the full lore.

    Args:
        ctx: Tool context with world info

    Returns:
        Tool function for drafting world
    """

    @tool(
        "draft_world",
        "Create a lightweight world draft with genre, theme, and lore summary. "
        "Call this FIRST to unblock sub-agents. They will use this context to "
        "create thematically consistent content while you refine the full lore.",
        DraftWorldInput.model_json_schema(),
    )
    async def draft_world(args: dict[str, Any]):
        """Create lightweight world draft to unblock sub-agents."""
        try:
            validated = DraftWorldInput.model_validate(args)
            world_name = ctx.require_world_name()

            # Ensure world directory exists
            config = WorldService.ensure_world_exists(world_name)

            # Save genre and theme to world config
            config.genre = validated.genre
            config.theme = validated.theme
            WorldService.save_world_config(world_name, config)

            # Save lore summary as initial lore.md
            lore_content = f"# World Lore\n\n{validated.lore_summary}"
            WorldService.save_lore(world_name, lore_content)

            logger.info(f"✅ World draft created for '{world_name}'")
            logger.info(f"Genre: {validated.genre}, Theme: {validated.theme}")

            return {
                "content": [
                    {
                        "type": "text",
                        "text": f"World draft created. Genre: {validated.genre}, Theme: {validated.theme}. "
                        f"Sub-agents can now start with this context.",
                    }
                ],
            }

        except Exception as e:
            logger.error(f"Failed to create world draft: {e}", exc_info=True)
            return {
                "content": [{"type": "text", "text": f"Error creating world draft: {e}"}],
                "is_error": True,
            }

    return draft_world


# =============================================================================
# Persist World Tool (Comprehensive - consolidates everything)
# =============================================================================


def _create_persist_world_tool(ctx: ToolContext):
    """
    Create the persist_world tool for Onboarding_Manager.

    This tool consolidates full lore with stat system and player state.
    Called AFTER sub-agents have started with draft context.
    Overwrites the draft lore with the full version.

    Args:
        ctx: Tool context with world info

    Returns:
        Tool function for persisting world
    """

    @tool(
        "persist_world",
        "Persist comprehensive world data: full lore (overwrites draft) + stat system + player state. "
        "Call this AFTER sub-agents have started with draft_world context.",
        PersistWorldInput.model_json_schema(),
    )
    async def persist_world(args: dict[str, Any]):
        """Persist comprehensive world data (full lore + stats)."""
        from services.player_service import PlayerService

        try:
            validated = PersistWorldInput.model_validate(args)
            world_name = ctx.require_world_name()

            # Ensure world directory exists
            WorldService.ensure_world_exists(world_name)

            # Save stat definitions
            stat_definitions = {
                "stats": [stat.model_dump() for stat in validated.stat_system.stats],
                "derived": validated.stat_system.derived or [],
            }
            PlayerService.save_stat_definitions(world_name, stat_definitions)
            logger.info(f"Saved {len(validated.stat_system.stats)} stat definitions")

            # Initialize player state
            state = PlayerService.load_player_state(world_name)
            if state:
                # Set default stats
                default_stats = {stat.name: stat.default for stat in validated.stat_system.stats}
                # Apply any overrides
                if validated.initial_stats:
                    default_stats.update(validated.initial_stats)
                state.stats = default_stats
                PlayerService.save_player_state(world_name, state)
                logger.info(f"Initialized player state with {len(state.stats)} stats")

            # Write full lore (overwrites draft, preserving world notes if they exist)
            existing_lore = WorldService.load_lore(world_name)
            world_notes_marker = "\n\n---\n## World Notes"
            if world_notes_marker in existing_lore:
                # Preserve the World Notes section
                notes_index = existing_lore.find(world_notes_marker)
                world_notes_section = existing_lore[notes_index:]
                new_lore = f"# World Lore\n\n{validated.lore}{world_notes_section}"
            else:
                new_lore = f"# World Lore\n\n{validated.lore}"

            # Add world notes if provided
            if validated.world_notes:
                world_notes = validated.world_notes.replace("\\n", "\n")
                new_lore = f"{new_lore}\n\n---\n## World Notes\n{world_notes}"

            WorldService.save_lore(world_name, new_lore)
            logger.info(f"✅ World persisted for '{world_name}'")

            return {
                "content": [
                    {
                        "type": "text",
                        "text": f"World persisted successfully. Stats: {len(validated.stat_system.stats)}, "
                        f"Lore: {len(validated.lore)} characters",
                    }
                ],
            }

        except Exception as e:
            logger.error(f"Failed to persist world: {e}", exc_info=True)
            return {
                "content": [{"type": "text", "text": f"Error persisting world: {e}"}],
                "is_error": True,
            }

    return persist_world

File: backend/src/sdk/handlers/onboarding-tools.ts
import { addCharacterToLocation, getLocationByName } from '@/crud/locations'
import { getAgentByName } from '@/crud/agents'
import type { LocationStorage } from '@/services/location-storage'
import type { PlayerService } from '@/services/player-service'
import type { WorldService } from '@/services/world-service'
import { WorldResetService } from '@/services/world-reset-service'
import { getLoreGuidelinesConfig } from '@/sdk/loaders/yaml-config'
import { formatTemplate } from '@/sdk/tools/definitions'
import {
  completeTool,
  draftWorldTool,
  persistWorldTool,
  readLoreGuidelinesTool,
  setWorldSettingsTool,
  worldStatusTool,
} from '@/sdk/tools/onboarding'
import {
  NAMING_STYLE_KEY,
  STYLE_NOTES_KEY,
  renderWorldSettingsBrief,
  toWorldSettings,
} from '@/domain/world-settings'
import { resolveTool } from '@/sdk/tools/registry'
import { getLogger } from '@/infrastructure/logging/logger'
import type { AgentFilesystemService } from '@/services/agent-filesystem-service'
import type { ItemService } from '@/services/item-service'
import { composeLore, listAdditionTitles, splitLore } from './lore-sections'
import { tool, requireWorldName, toolError, toolSuccess, type SdkTool, type ToolContext } from './context'

/**
 * World initialisation. `draft_world` is re-callable and merges, so the manager
 * can shape the world while the interview is still running and dispatch
 * designers between turns rather than in one batch at the end; `world_status`
 * is what it reads to avoid building the same place twice. Both writers go
 * through `lore-sections.ts`, because the lore file is no longer the manager's
 * alone — the designers write into it too (see `lore-tools.ts`).
 *
 * `set_world_settings` comes before both. It is the only writer of the world's
 * ground rules — language, player name, naming and style conventions — and what
 * it stores is rendered into every design sub-agent's prompt by
 * `sdk/agent/subagent-definitions.ts`, which is what stops a Korean world from
 * growing English items.
 *
 * Beyond flipping the phase, `complete` captures the initial-state snapshot the
 * reset feature restores to — it has to be taken after the item designer has
 * run, or "reset this world" hands the player an inventory they never had.
 */

const logger = getLogger('OnboardingTools')

export interface OnboardingDeps {
  worlds: WorldService
  players: PlayerService
  locations: LocationStorage
  reset: WorldResetService
  /** Needed to place onboarding-created NPCs; without it `complete` still works. */
  agentFiles?: AgentFilesystemService
  /** Only `world_status` reads it, and it degrades to "not reported" without it. */
  items?: ItemService
}

/** The lore guidelines text, from `lore_guidelines.yaml`'s active version. */
function loadLoreGuidelines(): string {
  const config = getLoreGuidelinesConfig()
  const activeVersion = typeof config.active_version === 'string' ? config.active_version : 'v1'
  const version = config[activeVersion]
  if (typeof version !== 'object' || version === null) return ''
  const template = (version as Record<string, unknown>).template
  return typeof template === 'string' ? template : ''
}

export function createOnboardingTools(ctx: ToolContext, deps: OnboardingDeps): SdkTool[] {
  const worldName = requireWorldName(ctx)
  const tools: SdkTool[] = []

  const settingsDef = resolveTool(setWorldSettingsTool.name, ctx.groupName)
  if (settingsDef) {
    tools.push(
      tool(
        setWorldSettingsTool.name,
        settingsDef.description,
        setWorldSettingsTool.inputSchema,
        async (args) => {
          if (
            args.language === null &&
            args.player_name === null &&
            args.naming_style === null &&
            args.style_notes === null
          ) {
            return toolError(
              'set_world_settings needs at least one of language, player_name, ' +
                'naming_style or style_notes. Call world_status to see what is registered.',
            )
          }

          try {
            // Merge, like `draft_world`: a later call fixing the naming convention
            // must not blank the language the world was created with.
            const config = deps.worlds.ensureWorldExists(worldName)
            if (args.language !== null) config.language = args.language
            if (args.player_name !== null) config.userName = args.player_name
            // Copied rather than mutated in place: the config came out of the
            // service's mtime cache, and the entry is shared with every other
            // reader until the save invalidates it.
            const settings = { ...config.settings }
            if (args.naming_style !== null) settings[NAMING_STYLE_KEY] = args.naming_style
            if (args.style_notes !== null) settings[STYLE_NOTES_KEY] = args.style_notes
            config.settings = settings
            deps.worlds.saveWorldConfig(worldName, config)

            // Echoed back verbatim: this is the same text the designers receive,
            // so the manager can see what it just committed them to.
            return toolSuccess(
              formatTemplate(settingsDef.response, {
                settings: renderWorldSettingsBrief(toWorldSettings(config)),
              }),
            )
          } catch (error) {
            logger.error(`Failed to register world settings: ${String(error)}`)
            return toolError(`Error registering world settings: ${String(error)}`)
          }
        },
      ),
    )
  }

  const loreDef = resolveTool(readLoreGuidelinesTool.name, ctx.groupName)
  if (loreDef) {
    // Read once, at build time. The file is hot-reloaded through the mtime cache
    // anyway, and a turn that reads it twice must not see two different texts.
    const guidelines = loadLoreGuidelines()
    tools.push(
      tool(
        readLoreGuidelinesTool.name,
        loreDef.description,
        readLoreGuidelinesTool.inputSchema,
        async () => toolSuccess(guidelines),
      ),
    )
  }

  const statusDef = resolveTool(worldStatusTool.name, ctx.groupName)
  if (statusDef) {
    tools.push(
      tool(
        worldStatusTool.name,
        statusDef.description,
        worldStatusTool.inputSchema,
        // Read at call time, not at build time: the point of this tool is to
        // report what the designers dispatched earlier in the turn have written.
        async () => toolSuccess(buildWorldStatus(deps, worldName)),
      ),
    )
  }

  // Gated like the others: an unconditional tool cannot be turned off from a
  // group config.
  const draftDef = resolveTool(draftWorldTool.name, ctx.groupName)
  if (draftDef) {
    tools.push(
      tool(
        draftWorldTool.name,
        draftDef.description,
        draftWorldTool.inputSchema,
        async (args) => {
          if (args.genre === null && args.theme === null && args.lore_summary === null) {
            return toolError(
              'draft_world needs at least one of genre, theme or lore_summary. ' +
                'Call world_status to see what the draft currently holds.',
            )
          }

          try {
            // Merge, not replace: a later call refining the theme must not blank
            // the genre the player settled two turns ago.
            const config = deps.worlds.ensureWorldExists(worldName)
            config.genre = args.genre ?? config.genre
            config.theme = args.theme ?? config.theme
            deps.worlds.saveWorldConfig(worldName, config)

            // The body is overwritten wholesale by `persist_world` later; it
            // exists only so designers running in the background have a world to
            // design for. Their own contributions live in a separate region and
            // survive both this write and that one.
            if (args.lore_summary !== null) {
              const sections = splitLore(deps.worlds.loadLore(worldName))
              deps.worlds.saveLore(
                worldName,
                composeLore({ ...sections, body: `# World Lore\n\n${args.lore_summary}` }),
              )
            }

            return toolSuccess(
              `World draft updated. Genre: ${config.genre ?? 'unset'}, ` +
                `Theme: ${config.theme ?? 'unset'}. ` +
                `Sub-agents can now start with this context.`,
            )
          } catch (error) {
            logger.error(`Failed to create world draft: ${String(error)}`)
            return toolError(`Error creating world draft: ${String(error)}`)
          }
        },
      ),
    )
  }

  const persistDef = resolveTool(persistWorldTool.name, ctx.groupName)
  if (persistDef) {
    tools.push(
      tool(
        persistWorldTool.name,
        persistDef.description,
        persistWorldTool.inputSchema,
        async (args) => {
          try {
            deps.worlds.ensureWorldExists(worldName)

            const stats = args.stat_system.stats
            deps.players.saveStatDefinitions(worldName, {
              stats: stats.map((stat) => ({ ...stat })),
              derived: args.stat_system.derived ?? [],
            })

            const state = deps.players.loadPlayerState(worldName)
            if (state) {
              const values: Record<string, number> = {}
              for (const stat of stats) values[stat.name] = stat.default
              // Overrides layer *over* the declared defaults rather than
              // replacing the map, so a world can start the player at half health
              // without having to restate every other stat.
              Object.assign(values, args.initial_stats ?? {})
              state.stats = values
              deps.players.savePlayerState(worldName, state)
            }

            // The draft body is replaced; the designers' lore sections and any
            // existing world notes are not the Onboarding Manager's to discard.
            const sections = splitLore(deps.worlds.loadLore(worldName))
            let notes = sections.notes
            if (args.world_notes) {
              // The model emits literal backslash-n inside a YAML-ish string
              // often enough to be worth un-escaping by hand.
              const added = args.world_notes.replaceAll('\\n', '\n')
              notes = notes ? `${notes}\n\n${added}` : added
            }

            deps.worlds.saveLore(
              worldName,
              composeLore({ ...sections, body: `# World Lore\n\n${args.lore}`, notes }),
            )

            return toolSuccess(
              `World persisted successfully. Stats: ${stats.length}, ` +
                `Lore: ${args.lore.length} characters`,
            )
          } catch (error) {
            logger.error(`Failed to persist world: ${String(error)}`)
            return toolError(`Error persisting world: ${String(error)}`)
          }
        },
      ),
    )
  }

  const completeDef = resolveTool(completeTool.name, ctx.groupName)
  if (completeDef) {
    tools.push(
      tool(
        completeTool.name,
        completeDef.description,
        completeTool.inputSchema,
        async (args) => {
          const {
            player_name: playerName,
            starting_location: startingLocation,
            starting_hour: startingHour,
          } = args

          try {
            // Checked against the *filesystem* index and by exact folder name.
            // A display name here would pass the model's own sense check and then
            // leave the player standing at a location that does not resolve.
            const fsLocations = deps.locations.loadAllLocations(worldName)
            if (!Object.hasOwn(fsLocations, startingLocation)) {
              const available = Object.keys(fsLocations)
              return toolError(
                `Starting location '${startingLocation}' not found. ` +
                  `Available locations: ${available.length > 0 ? available.join(', ') : 'none'}. ` +
                  `Make sure location_designer created this location first.`,
              )
            }

            const config = deps.worlds.ensureWorldExists(worldName)
            config.userName = playerName
            // *Pending*, not applied: the phase flips once this turn ends, so the
            // agent finishing its response is not reading a world that changed
            // underneath it mid-turn.
            config.pendingPhase = 'active'
            deps.worlds.saveWorldConfig(worldName, config)

            const playerState = deps.players.loadPlayerState(worldName)
            if (playerState) {
              playerState.currentLocation = startingLocation
              playerState.gameTime = { hour: startingHour, minute: 0, day: 1 }
              deps.players.savePlayerState(worldName, playerState)

              deps.reset.saveInitialState(
                worldName,
                WorldResetService.createInitialStateSnapshot({
                  startingLocation,
                  initialStats: playerState.stats,
                  initialInventory: playerState.inventory,
                  initialGameTime: playerState.gameTime,
                }),
              )
            }

            if (deps.agentFiles && ctx.worldId !== undefined) {
              try {
                addWorldAgentsToInitialLocation(ctx, deps.agentFiles, worldName, startingLocation)
              } catch (error) {
                logger.warning(`Failed to add agents to initial location: ${String(error)}`)
              }
            }
          } catch (error) {
            logger.error(`Failed to complete onboarding: ${String(error)}`)
            return toolError(`Error completing onboarding: ${String(error)}`)
          }

          return toolSuccess(
            formatTemplate(completeDef.response, {
              player_name: playerName,
              starting_location: startingLocation,
              starting_hour: startingHour,
            }),
          )
        },
      ),
    )
  }

  return tools
}

/**
 * A plain-text inventory of the world so far. Deliberately not a JSON dump: the
 * caller is a model deciding what to build next, and a missing section has to
 * read as "nothing here yet" rather than as an empty array it might skim past.
 * Every lookup is filesystem-first, so it also reflects a designer that finished
 * moments ago in the same turn.
 */
function buildWorldStatus(deps: OnboardingDeps, worldName: string): string {
  const lines: string[] = [`World: ${worldName}`]

  const config = deps.worlds.loadWorldConfig(worldName)
  lines.push(`Phase: ${config?.pendingPhase ?? config?.phase ?? 'onboarding'}`)
  lines.push(`Genre: ${config?.genre ?? '(not set)'}`)
  lines.push(`Theme: ${config?.theme ?? '(not set)'}`)
  lines.push(`Player name: ${config?.userName ?? '(not set — set_world_settings sets it)'}`)

  // Reported because they are what the designers are being handed: a world
  // whose language reads wrong here is producing content in that language.
  const settings = config ? toWorldSettings(config) : null
  lines.push(`Language: ${settings?.language ?? 'en'} (set_world_settings changes it)`)
  lines.push(`Naming convention: ${settings?.namingStyle ?? '(not registered)'}`)
  lines.push(`Style notes: ${settings?.styleNotes ?? '(not registered)'}`)

  const sections = splitLore(deps.worlds.loadLore(worldName))
  lines.push(`Lore body: ${sections.body.length} characters`)
  const titles = listAdditionTitles(sections.additions)
  lines.push(
    titles.length > 0
      ? `Lore sections from designers (${titles.length}): ${titles.join(', ')}`
      : 'Lore sections from designers: none yet',
  )

  // `persist_world` is the only writer of stats.json, so an empty catalog is
  // also the answer to "have I persisted the world yet".
  const stats = deps.players.loadStatDefinitions(worldName).stats ?? []
  lines.push(
    stats.length > 0
      ? `Stat system (${stats.length}): ${stats.map((stat) => stat.name).join(', ')}`
      : 'Stat system: not defined yet — persist_world writes it',
  )

  const state = deps.players.loadPlayerState(worldName)
  const locations = deps.locations.loadAllLocations(worldName)
  const locationNames = Object.keys(locations)
  lines.push(
    locationNames.length > 0
      ? `Locations (${locationNames.length}): ${locationNames
          .map((name) => {
            const display = locations[name]?.displayName ?? name
            const here = state?.currentLocation === name ? ' [starting]' : ''
            return `${name} (${display})${here}`
          })
          .join(', ')}`
      : 'Locations: none yet — dispatch location_designer',
  )

  if (deps.agentFiles) {
    const characters = deps.agentFiles.listWorldAgents(worldName)
    lines.push(
      characters.length > 0
        ? `Characters (${characters.length}): ${characters.join(', ')}`
        : 'Characters: none yet — dispatch character_designer',
    )
  }

  if (deps.items) {
    const items = deps.items.getAllItemsInWorld(worldName)
    lines.push(
      items.length > 0
        ? `Item templates (${items.length}): ${items.map((item) => item.itemId).join(', ')}`
        : 'Item templates: none yet — dispatch item_designer',
    )
  }

  return lines.join('\n')
}

/**
 * Place every NPC onboarding created at the starting location — they were
 * created against the *onboarding room*, which is an interview and not a place.
 * Per-agent failures are skipped: one unplaceable NPC must not cost the rest.
 */
function addWorldAgentsToInitialLocation(
  ctx: ToolContext,
  agentFiles: AgentFilesystemService,
  worldName: string,
  initialLocationName: string,
): number {
  const db = ctx.getDb()
  const agentNames = agentFiles.listWorldAgents(worldName)
  if (agentNames.length === 0) return 0

  const location = getLocationByName(db, ctx.worldId!, initialLocationName)
  if (!location) {
    logger.warning(`Initial location '${initialLocationName}' not found in database`)
    return 0
  }

  let added = 0
  for (const agentName of agentNames) {
    try {
      const agent = getAgentByName(db, agentName)
      if (!agent) {
        logger.warning(`Agent '${agentName}' not found in database`)
        continue
      }
      addCharacterToLocation(db, agent.id, location.id)
      added += 1
    } catch (error) {
      logger.warning(`Failed to add agent '${agentName}' to location: ${String(error)}`)
    }
  }

  logger.info(`Added ${added} agents to initial location`)
  return added
}
